README.md
<br>
<p align="center"><img align="center" alt="GitHub forks" width="800" src="./sakura-logo.png"></p>
<br>
<h2 align="center">Bringing you an easy, simple, spring like fresh start on Electron and Node.JS updaters.</h2>
<hr>
<p align="center">
<a href="https://github.com/ivanbogaeb/sakura/network"><img alt="GitHub forks" src="https://img.shields.io/github/forks/ivanbogaeb/sakura?style=flat-square"></a>
<a href="https://github.com/ivanbogaeb/sakura/stargazers"><img alt="GitHub stars" src="https://img.shields.io/github/stars/ivanbogaeb/sakura?style=flat-square"></a>
<a href="https://github.com/ivanbogaeb/sakura/issues"><img alt="GitHub issues" src="https://img.shields.io/github/issues/ivanbogaeb/sakura?style=flat-square"></a>
<a href="https://github.com/ivanbogaeb/sakura/blob/main/LICENSE"><img alt="GitHub license" src="https://img.shields.io/github/license/ivanbogaeb/sakura?style=flat-square"></a>
</p>

<br>

## What is Sakura?
Sakura is a minimal Electron and Node.JS auto-updater, offering developers high flexibility and modularity to work on from within their own code.

<br>

## Features:
- File checks.
- Splash screen. *(Electron only)*
- Automatic updates.
- Customizable alert *(Electron only)*
- Latest version available.
- Latest version downloader.
- Private and public repository access.

<br>

## To-do:
- [ ] Linux and iOS full support.
- [ ] Relaunch command on Node.JS apps.
    *(Currently Node.JS apps cannot execute themselves once updated)*

<br>

## Installation:

```powershell
npm install @ivanbogaeb/sakura
```

> Keep in mind that currently only Windows is fully supported.

<br>

## Example:

```javascript
const updated = async () => {
    const Sakura = require('@ivanbogaeb/sakura');
    try {
        let splashProperties = { // Create your own splash screen 👈
            show: true,
            frame: false,
            resizable: false,
            minimizable: false,
            maximizable: false,
            width: 400,
            height: 250,
            center: true,
            webPreferences: { // Sakura works under heavy security
                webSecurity: true,
                contextIsolation: true,
                preload: path.join(app.getAppPath(), './render/scripts/updater.js') // Sakura splash screen code!
            },
            icon: path.join(__dirname, './favicon.ico')
        };
        
        let sakura = new Sakura({
            gitUsername: 'ivanbogaeb',
            gitRepository: 'sakura',
            privateToken: 'ghp_loOVMTQMpXz0K****************************', // If you are using a private repo!
            installationDirectory: path.join(__dirname, '../'), // Your main app directory
            hasExecutable: true, // Toggle on if you are executing EXE files
            executableName: 'steroid.exe',
            isElectron: true, // Sakura works both for Electron and Node.JS
            currentVersion: require('./version.json'), // Load your current version from a file
        });

        // Let sakura auto-updater do the magic! Returns true when done!
        return await sakura.ready(splashProperties, path.join(app.getAppPath(), './render/html/updater.html'));
        
    } catch (error){
        console.log(error);
        return false; // Unable to update
    };
};

if (updated){
    // LOAD MAIN APP
};
```

<br>

## Reference:

- [Import](#import)
- [Settings](#settings)
- [Functions](#functions)
    - [Ready()](#ready-recommended)
    - [Check()](#check)
    - [Latest()](#latest)
    - [Download()](#download)
    - [Update()](#update)
    - [Splash()](#splash-electron-only)
    - [Close()](#close-electron-only)
    - [Messenger()](#messenger-electron-only-not-recommended-to-use)
- [Splash screen](#splash-screen)

### Import
You can simply import sakura as CommonJS module or ES, it doesn't matter!

- **CommonJS**:
    ```javascript
    const Sakura = require('@ivanbogaeb/sakura');
    ```
- **ES Module**:
    ```javascript
    import Sakura from '@ivanbogaeb/sakura';
    ```

### Settings
Sakura is completely customizable, in fact, you can use it's modules independently.

```javascript
let sakura = new Sakura({
    gitUsername: '', // Github username
    gitRepository: '', // Github repository
    privateToken: '', // ONLY FOR PRIVATE REPOSITORIES
    installationDirectory: path.join(__dirname, '../'), // Your main app directory
    hasExecutable: true/false, // Toggle on if you require to restart your application
    executableName: 'APP-NAME.exe', // Name of your main EXE file
    isElectron: true/false, // Sakura works both for Electron and Node.JS
    messages: {
        splash: 'Loading...', // Initial message
        fetching: 'Checking for updates...', // While looking for updates
        downloading: 'Downloading new update...', // When your app is downloading
        updating: 'Updating...', // While it's updating
        unableToUpdate: 'Unable to update, skipping...' // While it's updating
    },
    currentVersion: require('./version.json'), // Load your current version from a file, variable or string!
});
```

### Functions
With this module you have the options to automate the process or take control on your own!

#### **`ready()`** (Recommended)
Auto-updater function, it does all the process for you.
```javascript
let isReady = await sakura.ready(); // Returns true once the app has been updated or found it is up to date.
if (isReady){
    // Continue normal process
};
```

#### **`check()`**
Will check if you are running an update instance. If that's so, it will replace the old version with the new one and execute itself again.
```javascript
let isNotUpdate = await sakura.check(); // Returns true if it's not an update
if (isNotUpdate){
    // Continue normal process
};
```

#### **`latest()`**
Returns the latest version available.
```javascript
let latestVersion = await sakura.latest(); // Returns version string
```

#### **`download()`**
Creates an "Updates" folder inside your main application and downloads the latest version.
> Must be used along [latest()](#latest). This is because Sakura has a micro internal cache where it stores the current version you are working on and also the latest, including a download URL.
```javascript
await sakura.latest(); // Returns version string
let download = await sakura.download();
if (download){
    // Do something after it has been downloaded
};
```

#### **`update()`**
Updates your main application based on the latest version you downloaded.
> Must be used along [latest()](#latest) and [download()](#download).
```javascript
await sakura.latest(); // Returns version string
let download = await sakura.download();
if (download){
    await sakura.update();
};
```

#### **`splash()`** (Electron only)
Executes a splash screen of your choice.
```javascript
let splash = await sakura.splash(windowProperties, HTMLFile);
```

#### **`close()`** (Electron only)
Closes the splash screen.
```javascript
await sakura.close();
```

#### **`messenger()`** (Electron only) (Not recommended to use)
Talks to the splash screen by sending information throught actions.

**Types**:
- `activate`: Window you want to send information to throught [webContents](https://www.electronjs.org/es/docs/latest/api/web-contents).
    > Must be used with `window` type of Payload.
- `message`: Sends information to the splash screen.

**Payload**:
- `window`: [BrowserWindow](https://www.electronjs.org/es/docs/latest/api/browser-window) you want to connect to display updates information.
- `message`: This payload has four properties embed: `type`, `text`, `loader` and `percentage`.

**Message payload**:
```javascript
{
    type: 0, // Type 0 for messages, type 1 for downloads
    text: "This is an important updater message",
    loader: {
        active: false, // You can set a loader bar on and off
        data: 0, // Progress in kb
        total: 0, // Total size in kb
    }
}
```
```javascript
let message = await sakura.messenger({
    type: 'message',
    payload: {type: 0, text: "This is an important updater message", loader: {active: false, data: 0, total: 0}}
}); // Prints on the splash screen the text!

let message = await sakura.messenger({
    type: 'message',
    payload: {type: 1, text: "Downloading update...", loader: {active: true, data: 900, total: 1000}}
}); // Prints on the splash screen the text, activates the percentage bar to 90% completed.
```

### Splash Screen
Sakura communicates straight to the splash screen using [webContents](https://www.electronjs.org/es/docs/latest/api/web-contents). This way, you will be able to customize where you want updates information, how and which window.

- Create your own HTML, CSS and JavaScript files.
- Inside your JavaScript file, you will have to implement this code:
    ```javascript
    const { ipcRenderer } = require('electron');
    window.onload = async () => {
        ipcRenderer.on('SakuraUpdater', (event, message) => {
            /*
                Message contains:
                {
                    text: "Message you sent from sakura",
                    loader: {
                        active: true/false,
                        data: 0,
                        total: 0
                        percentage: 0 to 100
                    }
                }

            */
            document.getElementById('download-label').innerText = message.text; // Sakura message updates the download label text
            if (message.loader.active){
                document.getElementById('bar-progress').style.width = message.loader.percentage + '%'; // Progress bar using CSS properties
            };
        });
    };
    ```
- Setup your splash screen settings:
    ```javascript
    let splashScreenProperties = {
        show: true,
        frame: false,
        resizable: false,
        minimizable: false,
        maximizable: false,
        width: 400,
        height: 250,
        center: true,
        webPreferences: {
            webSecurity: true, // Always true
            contextIsolation: true, // Always isolate your Window from the rest
            preload: path.join(app.getAppPath(), './render/scripts/updater.js') // Preload your main JS file with required modules
        },
        icon: path.join(__dirname, '../favicon.ico')
    };
    ```
- Create a splash screen executing [Splash()](#splash-electron-only).

<br>

## Security notes:

The main reason why Sakura runs only throught [webContents](https://www.electronjs.org/es/docs/latest/api/web-contents) is to prevent malicious code injection, malicious activity, bugs or even poor performance or communication between the splash screen and Sakura. It has been noted that allowing Node modules to be executed inside a Window in Electron is not a good practice, and I took it seriously while developing this module.

<br>

## Changelog:
- 24 July 2022 - Bug fix - Version 1.0.2
- 24 July 2022 - First "Release" - Version 1.0.0
- **[Read more...](./CHANGELOG.md)**

<br>

## Credits:
- Inspired by **[UAUP-JS](https://github.com/DcmanProductions/UAUP-JS)**.

<br>

## License:
**[CC0 1.0 Universal](./LICENSE)**


CHANGELOG.md
# 1.0.4 / 16-09-2022
Due to some issues with how Windows handles programs on startup, this patch was brought.

**Changelog**:
- New message in case of update error has been added.
- All modules in charge of connecting to Github have been rewritten.
- Now Sakura supports errors and issues related to not being able to connect to Github.

# 1.0.2 / 24-07-2022

First bug fix, damn...

**Changelog**:
- sakura.ts has been modified due to an import error.

# 1.0.0 / 24-07-2022

First release, nothing fancy!

**Changelog**:
- None.


src/sakura.ts
interface ISakura {
    gitUsername: string,
    gitRepository: string,
    privateToken: string,
    installationDirectory: string,
    hasExecutable: boolean,
    executableName: string,
    isElectron: boolean,
    messages: {
        splash: string,
        fetching: string,
        downloading: string,
        updating: string
    },
    currentVersion: string
};

/**
* Sakura is a minimal Electron and Node.JS auto-updater, offering developers high flexibility and modularity to work on from within their own code.
* @example
* ```js
    const updated = async () => {
        const Sakura = require('@ivanbogaeb/sakura');
        try {
            let splashProperties = { // Create your own splash screen 👈
                show: true,
                frame: false,
                resizable: false,
                minimizable: false,
                maximizable: false,
                width: 400,
                height: 250,
                center: true,
                webPreferences: { // Sakura works under heavy security
                    webSecurity: true,
                    contextIsolation: true,
                    preload: path.join(app.getAppPath(), './render/scripts/updater.js') // Sakura splash screen code!
                },
                icon: path.join(__dirname, './favicon.ico')
            };
            
            let sakura = new Sakura({
                gitUsername: 'ivanbogaeb',
                gitRepository: 'sakura',
                privateToken: 'ghp_loOVMTQMpXz0K****************************', // If you are using a private repo!
                installationDirectory: path.join(__dirname, '../'), // Your main app directory
                hasExecutable: true, // Toggle on if you are executing EXE files
                executableName: 'SakuraApp.exe',
                isElectron: true, // Sakura works both for Electron and Node.JS
                currentVersion: require('./version.json'), // Load your current version from a file or text string
            });

            // Let sakura auto-updater do the magic! Returns true when done!
            return await sakura.ready(splashProperties, path.join(app.getAppPath(), './render/html/updater.html'));
            
        } catch (error){
            console.log(error);
            return false; // Unable to update
        };
    };

    if (updated){
        // LOAD MAIN APP
    };
* ```
*/
class Sakura implements ISakura {

    // Settings
    /**
     * Github username to fetch repos
     * @example
     * ```js
     * sakura.gitUsername = 'ivanbogaeb';
     * ```
    */
    gitUsername: string;
    /**
     * Github repository of your app
     * @example
     * ```js
     * sakura.gitRepository = 'sakura';
     * ```
    */
    gitRepository: string;
    /**
     * Your private token, here you can get yours: https://github.com/settings/tokens (DON'T USE THIS IN PRODUCTION)
     * @example
     * ```js
     * sakura.gitRepository = 'ghp_loOVMTQMpXz0K****************************';
     * ```
    */
    privateToken: string;
    /**
     * Main app installation directory (Just in case if you are not using electron)
     * @example
     * ```js
     * sakura.installationDirectory = path.join(__dirname, '../');
     * ```
    */
    installationDirectory: string;
    /**
     * If you need to also update an EXE file, use this function to halt them
     * @example
     * ```js
     * sakura.hasExecutable = true;
     * ```
    */
    hasExecutable: boolean;
    /**
     * Main EXE file that needs to be executed
     * @example
     * ```js
     * sakura.executableName = 'SakuraApp.exe';
     * ```
    */
    executableName: string;
    /**
     * If you are running an Electron application!
     * @example
     * ```js
     * sakura.isElectron = true;
     * ```
    */
    isElectron: boolean;
    /**
     * // Messages to be displayed on certain actions
    */
    messages: {
        splash: string;
        fetching: string;
        downloading: string;
        updating: string;
    };
    /**
     * // Current version of your program
     * @example
     * ```js
     * sakura.currentVersion = '1.0.0';
     * ```
    */
    currentVersion: string;
    /**
     * Will check if you are running an update instance. If that's so, it will replace the old version with the new one and execute itself again.
     * @example
    ```javascript
        let isNotUpdate = await sakura.check(); // Returns true if it's not an update
        if (isNotUpdate){
            // Continue normal process
        };
    ```
    */
    check: () => Promise<true>;
    /**
     * Returns the latest version available.
     * @example
    ```javascript
        let latestVersion = await sakura.latest(); // Returns version string
    ```
     * @url https://adasdas.com 
    */
    latest: () => Promise<any>;
    /**
     * Creates an "Updates" folder inside your main application and downloads the latest version.
     * @example
     * ```js
     *  await sakura.latest(); // Returns version string
        let download = await sakura.download();
        if (download){
            // Do something after it has been downloaded
        };
     * ```
    */
    download: () => Promise<boolean>;
    /**
     * Updates your main application based on the latest version you downloaded.
     * @example
     * ```js
     *  await sakura.latest(); // Returns version string
        let download = await sakura.download();
        if (download){
            await sakura.update();
        };
     * ```
    */
    update: () => Promise<any>;
    /**
     * Executes a splash screen of your choice.
     * @example
     * ```js
     * let splash = await sakura.splash(windowProperties, HTMLFile);
     * ```
    */
    splash: (windowProperties: object, HTMLFile: string) => Promise<any>;
    /**
     * Closes the splash screen.
     * @example
     * ```js
     * await sakura.close();
     * ```
    */
    close: () => void;
    /**
     * Talks to the splash screen by sending information throught actions.
     * @example
     * ```js
     *  let message = await sakura.messenger({
            type: 'message',
            payload: {type: 0, text: "This is an important updater message", loader: {active: false, data: 0, total: 0}}
        }); // Prints on the splash screen the text!

        let message = await sakura.messenger({
            type: 'message',
            payload: {type: 1, text: "Downloading update...", loader: {active: true, data: 900, total: 1000}}
        }); // Prints on the splash screen the text, activates the percentage bar to 90% completed.
     * ```
    */
    messenger: (action: object) => Promise<any>;
    /**
     * Auto-updater function, it does all the process for you.
     * @example
     * ```js
     *  let isReady = await sakura.ready(); // Returns true once the app has been updated or found it is up to date.
        if (isReady){
            // Continue normal process
        };
     * ```
    */
    ready: (windowProperties: object, HTMLFile: string) => Promise<any>;

    constructor({
        gitUsername = '',
        gitRepository = '',
        privateToken = '',
        installationDirectory = '',
        hasExecutable = true,
        executableName = '',
        isElectron = true,
        messages = {
            splash: 'Loading...',
            fetching: 'Checking for updates...',
            downloading: 'Downloading new update...',
            updating: 'Updating...',
            unableToUpdate: 'Unable to update, skipping...'
        },
        currentVersion = ''
    }){
        const fs = require('fs');
        const path = require('path');
        const AdmZip = require('adm-zip');
        const axios = require('axios').default;
        const { spawn } = require('child_process');
        const { compare } = require("compare-versions");
        
        // Imports
        const timer = require('./modules/timer');
        const check = require('./modules/check'); // Update checker, if we are running an update, it will instantly replace the old app with the new version.
        const ready = require('./modules/ready'); // Automatic update check, installation and run.
        const latest = require('./modules/latest'); // Get latest version available.
        const update = require('./modules/update'); // Extracts the downloaded app, runs it and closes the old version.
        const download = require('./modules/download'); // Downloads the latest version if available.
        const {windowMessenger, windowCreator, windowClose} = require('./modules/window'); // Information handler that connects the Sakura Browser Window and Application safely.

        // Settings
        this.gitUsername = gitUsername;
        this.gitRepository = gitRepository;
        this.privateToken = privateToken;
        this.installationDirectory = installationDirectory;
        this.hasExecutable = hasExecutable;
        this.executableName = executableName;
        this.isElectron = isElectron;

        // Messages
        this.messages = messages;

        // Versioning
        this.currentVersion = currentVersion;
        let newVersion = ''; // This is exclusively for internal use
        let downloadURL = ''; // Download URL of your release

        // Functions
        this.check = async () => {
            return await check(path, spawn, timer, windowMessenger, this.isElectron, this.messages, this.installationDirectory, this.executableName);
        };
        this.latest = async () => {
            let data:any = await latest(axios, timer, windowMessenger, this.isElectron, this.messages, this.gitUsername, this.gitRepository, this.privateToken);
            if (data.hasOwnProperty('tagname')){
                newVersion = data.tagname;
                downloadURL = data.url;
                return data.tagname;
            } else {
                return currentVersion;
            }
        };
        this.download = async () => {
            return await download(fs, path, axios, windowMessenger, this.isElectron, this.messages, downloadURL, this.privateToken, this.installationDirectory, newVersion);
        };
        this.update = async () => {
            if (this.currentVersion !== newVersion){
                return await update(path, AdmZip, spawn, windowMessenger, this.isElectron, this.messages, this.installationDirectory, newVersion, this.hasExecutable, this.executableName);
            } else {
                return false;
            };
        };
        this.splash = async (windowProperties: object, HTMLFile: string) => {
            return await windowCreator(this.isElectron, windowProperties, HTMLFile);
        };
        this.close = () => {
            windowClose(this.isElectron);
        };
        this.messenger = async (action: object) => {
            await windowMessenger(this.isElectron, action);
        }
        this.ready = async (windowProperties, HTMLFile) => {
            return await ready(compare, windowProperties, HTMLFile, this.splash, this.check, this.latest, this.download, this.update, this.close, this.currentVersion);
        };
    };
};

module.exports = Sakura;